code/modules/clothing/spacesuits/rig.dm

```diff
@@ -97,639 +111,310 @@
 	max_heat_protection_temperature = SPACE_SUIT_MAX_HEAT_PROTECTION_TEMPERATURE
 	pressure_resistance = 200 * ONE_ATMOSPHERE
+	actions_types = list(/datum/action/item_action/toggle_rig_suit)
+
+	var/activated = FALSE
+	var/list/modules = list()
+	var/list/initial_modules = list(/obj/item/rig_module/health_readout)
+
 	var/obj/item/clothing/head/helmet/space/rig/H = null
-	var/head_type = /obj/item/clothing/head/helmet/space/rig
+	var/obj/item/clothing/gloves/G = null
+	var/obj/item/clothing/shoes/magboots/MB = null
+	var/obj/item/weapon/tank/T = null
 	var/obj/item/weapon/cell/cell = null
+
+	var/head_type = /obj/item/clothing/head/helmet/space/rig
+	var/boots_type =  null
+	var/gloves_type = null
+	var/tank_type = null
 	var/cell_type = /obj/item/weapon/cell/high //The cell_type we're actually using
-	var/list/modules = list()
-	actions_types = list(/datum/action/item_action/toggle_rig_helmet)
+
+	var/mob/living/carbon/human/wearer
 
 /obj/item/clothing/suit/space/rig/New()
 	..()
-	cell = new cell_type
-	H = new head_type
+	if(cell_type)
+		cell = new cell_type(src)
+	if(head_type)
+		H = new head_type(src)
+		H.canremove = FALSE
+		H.rig = src
+	if(gloves_type)
+		G = new gloves_type(src)
+		G.canremove = FALSE
+	if(boots_type)
+		MB = new boots_type(src)
+		MB.canremove = FALSE
+	if(tank_type)
+		T = new tank_type(src)
+	if(initial_modules && initial_modules.len)
+		for(var/path in initial_modules)
+			var/obj/item/rig_module/new_module = new path(src)
+			modules += new_module
+	if(!(src in processing_objects))
+		processing_objects.Add(src)
 
 /obj/item/clothing/suit/space/rig/Destroy()
-	qdel(cell)
-	cell = null
-	if(H && (H.loc == src || !H.loc))
-		qdel(H)
+	if(processing_objects.Find(src))
+		processing_objects.Remove(src)
+	for(var/obj/item/I in list(H,G,T,MB))
+		if(I && (I.loc == src || !I.loc))
+			qdel(I)
 	H = null
+	G = null
+	T = null
+	MB = null
 	for(var/obj/M in modules)
 		qdel(M)
-	modules.Cut()
+	modules = null
+	if(cell)
+		qdel(cell)
+	cell = null
+	wearer = null
+	if(processing_objects.Find(src))
+		processing_objects.Remove(src)
 	..()
 
 /obj/item/clothing/suit/space/rig/examine(mob/user)
 	..()
 	for(var/obj/item/rig_module/M in modules)
 		M.examine_addition(user)
 
+/obj/item/clothing/suit/space/rig/get_cell()
+	return cell
+
+/obj/item/clothing/suit/space/rig/equipped(mob/living/carbon/human/user, var/slot)
+	..()
+	wearer = user
+
 /obj/item/clothing/suit/space/rig/unequipped(mob/living/carbon/human/user, var/from_slot = null)
 	..()
 	if(from_slot == slot_wear_suit && istype(user))
-		deactivate_suit(user)
+		deactivate_suit(user, TRUE)
+	wearer = null
 
-/obj/item/clothing/suit/space/rig/proc/toggle_helmet(mob/living/carbon/human/user)
-	if(!user.is_wearing_item(src, slot_wear_suit))
+/obj/item/clothing/suit/space/rig/dropped()
+	..()
+	for(var/piece in list(HARDSUIT_HEADGEAR,HARDSUIT_GLOVES,HARDSUIT_BOOTS))
```

Comment:
If my understanding is correct, this list is created every single time you are calling the suit's `dropped()`, stored for the duration of the proc, and then deleted. I think it would make more sense to get us a constant list, either a global constant, or a static one.

Reply:
good point.